supabase/functions/process-transcripts/index.ts

```diff
@@ -0,0 +1,190 @@
+      }
+
+    const supabase = supabaseClient(req);
+    const openaiClient = new OpenAI({
+        apiKey: Deno.env.get('OPENAI_API_KEY')!,
+        });
+
+    try {
+        const { data, error } = await supabase
+            .from('records')
+            .select('raw_text, id')
+            .eq('processed', false)
+            .order('created_at', { ascending: true })
+            .limit(5);
+
+        if (error) {
+            console.error('Error fetching records:', error);
+            return new Response(
+                JSON.stringify({ message: 'Error fetching records.' }),
+                {
+                    headers: {
+                        ...corsHeaders,
+                        'Content-Type': 'application/json',
+                    },
+                    status: 500,
+                }
+            );
+        }
+
+        // Create a broader context
+        let concatenatedTranscripts: string;
+        if (data.length > 0) {
+            concatenatedTranscripts = data
+                .map((record: any) => record.raw_text)
+                .join(' ');
+        } else {
+            return new Response(
+                JSON.stringify({ message: 'No records found.' }),
+                {
+                    headers: {
+                        ...corsHeaders,
+                        'Content-Type': 'application/json',
+                    },
+                    status: 404,
+                }
+            );
+        }
+
+        const response = await openaiClient.chat.completions.create({
+            model: 'gpt-4-0125-preview',
+            messages: [
+                {
+                    role: 'system',
+                    content: `
+            These transcripts contain information about your user. 
+            Your task is to organize the information in a way that makes sense to you.
+            Your response must be in json format with the three following keys: "summary", "topics".
+            `,
+                },
+                {
+                    role: 'user',
+                    content: `${concatenatedTranscripts}\n\nGiven the information about the user, provide a summary, and the topics discussed.\n
+            *** Summary must be a brief overview of the transcript.\n\n
+            *** Topics must be a list of topics that were discussed in the transcript, include topics not mentioned but that relate to the topics discussed.\n\n
+             `,
+                },
+            ],
+            response_format: { type: 'json_object' },
+        });
+
+
+        const responseData = JSON.parse(
+            response.choices[0].message.content
+        );
+
+        const { summary, topics } = responseData;
+
+        // Create meta data for each indivual transcript
+        let transcript_and_metadata: Object = {};
+        for (const record of data) {
+            const response = await openaiClient.chat.completions.create({
```

Comment:
So if I understand it correctly, here you're regenerating summary, topics, 
 etc for prevs transcripts that were already in the database?
 
 This is the case because you've seperated the main func into two functions. and now you fetch the "working" memory from database instead of the http request.
 
 This should not be the case, as it will more likely to cause issues, be more expensive (as you already generated summaries for some of the trascript before, in prevs requests), etc.


Instead, you should have one edge function that:

1. Get the audio file from the request
2. transcribe it using Whisper
3. Fetch 5 prevs entries from database
4. send 1 request to gpt4 to generate metadata, summary, etc
5. Generate 2 embeddings: the new generated summary, and the -current- whisper text.

you shouldn't do it do it for prevs entries, as it should already exist 

Let me know if this comment confuses you, happy to jump on a call and walk this through :)

Reply:
The only reason I split the functions was because of state. I wanted to transcribe 5 recordings before starting to embed. Since serverless maintains no state this was my solution. I’ll have a refactor in today. Thanks for taking the time to look it over!

Reply:
Yeah I understand that, and I see the changes you made to still transcribe 5 recordings at a time.

That is interesting, and I'm not entirely sure what I am going to suggest is the more right approach, but: 

My thoughts were to process the new coming entry without buffering to 5 - basically doing the following:
new entry coming in -> fetch the last 5 -> process these 6 together in gpt4 -> save new data to the new entry.
Sometime like this:
1, 2, 3, [4 comes in] -> process
2, 3, 4, [5 comes in] -> process
3, 4, 5, [6 comes in] -> process
4, 5, 6, [7 comes in] -> process
5, 6, 7, [8 comes in] -> process
...

This will result in having each entry on the database having its own context, using the past 5 entries. This will basically create some kind of "chain" and data "overlap":
<img width="662" alt="image" src="https://github.com/adamcohenhillel/ADeus/assets/633840/22dab058-3a71-43a0-85ed-03239fb4b261">

where you have a chain all the way back to the first entry.

In "RAG" this is called "chunk overlap":
![image](https://github.com/adamcohenhillel/ADeus/assets/633840/a31ce4a5-07c6-46da-a549-12c12171a074)
see:
1. https://gustavo-espindola.medium.com/chunk-division-and-overlap-understanding-the-process-ade7eae1b2bd
2. https://dev.to/peterabel/what-chunk-size-and-chunk-overlap-should-you-use-4338


Not sure this is better, because it will be more "expensive" and "slow", but will give us a better way of understanding the data later, what do you think?